base/abstractarray.jl

```diff
@@ -1133,7 +1133,19 @@ parts can specialize this method to return the concatenation of the `dataids` of
 their component parts.  A typical definition for an array that wraps a parent is
 `Base.dataids(C::CustomArray) = dataids(C.parent)`.
 """
-dataids(A::AbstractArray) = (UInt(objectid(A)),)
+function dataids(A::AbstractArray)
+    @_inline_meta
+    if @generated
+        :(ids = tuple($([:(dataids(getfield(A, $i))...) for i in 1:fieldcount(A)]...)))
+    else
+        ids = _splatmap(dataids, ntuple(i -> getfield(A, i), Val(nfields(A))))
+    end
+    if isimmutable(A) || !isempty(ids)
```

Comment:
I think we also want to include `UInt(pointer_from_objref(A))` in the case where `!isimmutable(A)`, yeah?  I think the branching on `isempty(ids)` is kind of a red herring.

```julia
if isimmutable(A)
    return ids
else
    return (UInt(pointer_from_objref(A)), ids...)
end
```

E.g., a mutable array struct could derive some of its elements from primitive type fields.

Reply:
I've tried to make that argument above. In aliasing detection, we're only interested whether at least one dataid is equal. So for two distinct objects, including their PFOs only leads to a number of failing comparisons, but we still need to consider all other dataids, so this only adds work to do for no good. If the two objects are identical (so it's one and the same, really) all their dataids have an equal counterpart, so no need to add the PFO as yet another one. Unless there are no other dataids (`isempty`). Does that make that sense?

Reply:
Yup, totally, thanks for the explanation.  For some reason this feels surprisingly counter-intuitive to me, but of course the contained dataids would _also_ match if we have the same object.